ciao-controller/internal/datastore/datastore.go

```diff
@@ -681,11 +562,52 @@ func (ds *Datastore) ReleaseTenantIP(tenantID string, ip string) error {
 	ds.tenantsLock.Lock()
 
 	if ds.tenants[tenantID] != nil {
-		ds.tenants[tenantID].network[int(subnetInt)][int(ipBytes[3])] = false
+		delete(ds.tenants[tenantID].network[int(subnetInt)], int(ipBytes[3]))
+		subnets := ds.tenants[tenantID].subnets
+		network := ds.tenants[tenantID].network
+		i = int(subnetInt)
+
+		if len(network[i]) == 0 {
+			// delete the network map and the subnet
+			delete(ds.tenants[tenantID].network, i)
+
+			if len(subnets) > 1 {
+				ds.tenants[tenantID].subnets = append(subnets[:i], subnets[i+1:]...)
+			} else {
+				ds.tenants[tenantID].subnets = nil
+			}
+
+			removeSubnet = true
+		}
 	}
 
 	ds.tenantsLock.Unlock()
 
+	// must be called with tenants lock unlocked?
+	// should we wait 5 minutes here?
+	if removeSubnet && ds.tenants[tenantID].CNCIctrl != nil {
+		go func() {
+			time.Sleep(5 * time.Minute)
+			ds.tenantsLock.RLock()
+			subnets := ds.tenants[tenantID].subnets
+			for _, s := range subnets {
+				if s == i {
+					removeSubnet = false
+					break
+				}
+			}
+			ds.tenantsLock.RUnlock()
+
+			if removeSubnet {
+				// handle error?
+				err := ds.tenants[tenantID].CNCIctrl.RemoveSubnet(i)
+				if err != nil {
+					glog.Warningf("Error removing subnet: (%v)\n", err)
+				}
+			}
+		}()
```

Comment:
In general when you start a goroutine you need to handle tidying up (especially for a long lived one). For this one, create a context.WithCancel() and save the cancel function into a map on the datastore (or in the tenant struct.)

In the goroutine select {} on both a time.After and the Done() channel on the context. On the former do your deletion on the latter just return and terminate the goroutine.

In the datastore shutdown call all the cancel functions  which will safely terminate all the running goroutines. You can also call the cancel if you get a new instance for a pending closed subnet removing the need to even check if its in use.